src/lib.rs

```diff
@@ -263,22 +293,30 @@ fn request<U: IntoUrl>(u: U) -> Result<String> {
 }
 
 impl List {
-    fn append(&mut self, rule: &str, typ: Type) -> Result<()> {
-        rule.rsplit('.').next()
-            .ok_or(ErrorKind::InvalidRule(rule.into()).into())
-            .and_then(|tld| {
-                if tld.is_empty() {
-                    return Err(ErrorKind::InvalidRule(rule.into()).into());
-                }
-                Ok(tld)})
-            .and_then(|tld| {
-                self.rules.entry(tld.into()).or_insert(Vec::new())
-                    .push(Suffix {
-                        rule: rule.into(),
-                        typ: typ,
-                    });
-                Ok(())
-            })
+    fn append(&mut self, mut rule: &str, typ: Type) -> Result<()> {
+        let mut is_exception_rule = false;
+        if rule.starts_with("!") {
+            is_exception_rule = true;
+            rule = &rule[1..];
+        }
+
+        let mut current = &mut self.root;
+        for label in rule.rsplit('.') {
+            if label.is_empty() {
+                return Err(ErrorKind::InvalidRule(rule.into()).into());
+            }
+
+            let cur = current;
+            current = cur.children.entry(label.to_owned())
+                .or_insert(Box::new(ListNode::new()));
+        }
+
+        current.leaf = Some(ListLeaf::new(typ, is_exception_rule));
+
+        // to support all(), icann(), private()
+        self.all.push(Suffix {rule: rule.to_owned(), typ: typ});
```

Comment:
Since the `all()` method returns a reference to str, all strings must be kept inside. If we deprecate the `all()` method or change it to return a `String` type, we will not need to do this.